app/src/main/kotlin/compiler/intermediate_form/SpillsHandlingAllocation.kt

```diff
@@ -1,20 +1,150 @@
+                    registerAllocation,
+                    spilledRegistersColoring,
+                )
+
+                return Result(
+                    registerAllocation,
+                    newLinearProgram,
+                    (spilledRegistersColoring.values.maxOrNull() ?: 0u) * memoryUnitSize,
+                )
+            }
+
+            reservedRegistersNumber ++
+        }
+    }
+
+    data class SpilledInstruction(
+        val instruction: Instruction,
+        val spilledUsedRegisters: Set<Register>,
+        val spilledDefinedRegisters: Set<Register>,
+        val numberOfSpilledRegisters: Int = (spilledUsedRegisters + spilledDefinedRegisters).size,
+        val isSpilled: Boolean = (numberOfSpilledRegisters > 0),
+    )
+
+    private fun toSpilledInstruction(
+        instruction: Instruction,
+        spilledRegisters: Set<Register>,
+    ): SpilledInstruction = SpilledInstruction(
+        instruction,
+        spilledRegisters.intersect(instruction.regsUsed.toSet()),
+        spilledRegisters.intersect(instruction.regsDefined.toSet()),
+    )
+
+    private fun colorSpilledRegisters(
+        spilledRegisters: Set<Register>,
+        livenessGraphs: Liveness.LivenessGraphs,
+        allocator: Allocation,
+    ): Map<Register, ULong> {
+        val spillsPossibleColors = spilledRegisters.map { Register() }
+        val spillsColoring: Map<Register, Register> = allocator.allocateRegisters(
+            Liveness.LivenessGraphs(
+                Liveness.inducedSubgraph(livenessGraphs.interferenceGraph, spilledRegisters),
+                Liveness.inducedSubgraph(livenessGraphs.copyGraph, spilledRegisters),
+            ),
+            spillsPossibleColors,
+        ).allocatedRegisters
+        val spillsColors = spillsPossibleColors.filter { it in spillsColoring.values }
+        return spillsColoring.entries.associate { it.key to (spillsColors.indexOf(it.value) + 1).toULong() }
+    }
+
+    private fun allocateSpilledRegisters(
+        reservedRegisters: List<Register>,
+        partialAllocation: Map<Register, Register>,
+        spilledInstructions: List<SpilledInstruction>,
+    ): Map<Register, Register> {
+        val newAllocation = partialAllocation.toMutableMap()
+        spilledInstructions
+            .filter { it.isSpilled }
+            .map { it.spilledUsedRegisters + it.spilledDefinedRegisters }
+            .forEach { spilledRegisters -> spilledRegisters.withIndex().forEach { newAllocation[it.value] = reservedRegisters[it.index] } }
+        return newAllocation
+    }
+
+    private fun transformProgram(
+        linearProgram: List<Asmable>,
+        instructionsToSpilled: Map<Int, SpilledInstruction>,
+        allocatedRegisters: Map<Register, Register>,
+        spillsColoring: Map<Register, ULong>,
+    ): List<Asmable> {
+        val newLinearProgram = mutableListOf<Asmable>()
+
+        fun handleSpilledInstruction(spilledInstruction: SpilledInstruction) {
```

Comment:
Is moving in this function for sure correct?
Now we save each used register to memory before usage and override register with memory after definition.
Shouldn't we save to memory before definition and load from memory before usage

Reply:
Hmm, first, we load to registers to be used from memory (MoveRM), then we perform the instruction and finally, we save defined registers to memory (MoverMR). How can we save to memory before definition (when the thing is not defined yet)?

Reply:
Oh, I see now. `RM` means memory -> register and `MR` register -> memory. I thought it was the other way around.

Reply:
I hope I'm right
```
MoveRM(val reg_dest: Register, val mem_src: Addressing)
MoveMR(val mem_dest: Addressing, val reg_src: Register)
```

Reply:
Yeah, all good.